content/purgatorio/now-go-build/transfer-learning.md

```diff
@@ -1,5 +1,111 @@
+
+For example, you can fine-tune a pre-trained model from ImageNet on your much smaller dataset of images, in order to benefit the pre-training and obtain a series of almost-magical properties and results.
+
+To get a comprehensive overview of Why, What and How is used Transfer Learning, plus possible applications, read this awesome article:
+
+- [Transfer Learning - Machine Learning's Next Frontier](https://ruder.io/transfer-learning/index.html)
+
+
+#### Transfer Learning Magical Properties
+
+The theory of Transfer Learning suggests that when using a pre-trained model on a similar dataset:
+
+**You need less**:
+- Data: you need much smaller datasets to obtain the same (or better) model performances
+- Computational Power / Computational Time: empirical results have shown that pre-trained models reach top performances after much fewer epochs of training, that starting the training from scratch
+
+**You improve:**
+- Out-of-sample generalization: when predicting on test data you often obtain better accuracy, thus reducing overfitting (or whatever metric you're using)
+- Robustness: you make the model more "robust" to real-world low-quality data  
+
+So, **the first thing to do** every time you frame a new problem, is to ask yourself:
+
+**Can I leverage Transfer Learning in solving this problem?**
+
+
+## How Does Transfer Learning Work
+
+The practice of Transfer Learning allows reusing most of the parameters (weights) of a neural network previously trained on a problem similar to the one we have to solve, dwelling on the training only of the last layers that are usually those dedicated to the classification and/or regression of the features obtained with the previous layers.
+
+This allows us to obtain two key results:
+
+- reuse of the behavior of a network already trained to effectively **extract features** from input data
+- limit processing to a significantly **smaller number of parameters** (corresponding to the last layers)
+
+For example, if we were to classify apple varieties from an image, we could start by using a neural network already trained to classify images of planes, cars, dogs, cats, eggs, etc... This is because of the greater variety of training dataset ensures a better ability to extract features of various kinds from images.
+
+Of the _pre-trained neural network_, we would keep only the initial layers, going to redefine only the last layers of classification.
+
+The reused layers would be labeled as "read-only", in order to calculate only the parameters corresponding to the last layers, speeding up the training time, reducing the processing power required, and generally improving the accuracy.
+
+In general, the set of layers that are reused by a pre-trained network is called a backbone or feature-extractor.
+
+The practice of applying Transfer Learning theory to real-world models is called "fine-tuning".
```

Comment:
Also mention experimentation with removing _last n_ layers, by default (or by practice) the last layer is removed, but what if _last 2_ or _last 3_ layers are removed - this can help in cases, specific to use-cases.

Reply:
cool idea, added